d2/src/test/java/com/linkedin/d2/balancer/util/canary/TestBasicCanaryDistributionProviderImpl.java

```diff
@@ -183,28 +183,44 @@ public void testEdgeCases(D2CanaryDistributionStrategy strategy)
         CanaryDistributionProvider.Distribution.STABLE, "Invalid strategies should return stable");
   }
 
+  // Verify uniform distribution of the String built-in hashCode method, which theoretically should be guaranteed as it's used
+  // in Map with String keys. But we explicitly test it to verify its behavior.
   @Test
   public void testHashDistribution()
   {
     // verify hash results are almost uniformly distributed
-    int bucketCount = 100;
+    int bucketCount = 100; // since ramp is 100% based, we test with 100 buckets
     int attemptCount = 1000;
+    // count how many attempts are skewed
+    // (e.g: one bucket gets 13 will cause another bucket gets 7, the attempts skewed in this pair of buckets is ((13 - 10) + (10 - 7)) / 2 = 3
+    int skewedCount = 0;
     int[] hits = new int[bucketCount]; // to count hits in each bucket 0 ~ 99.
     String serviceName = "test-service";
     String hostName = "lor1-app";
     int bucketIdx;
-    int hashResult;
+    BasicCanaryDistributionProviderImpl impl = spy(new BasicCanaryDistributionProviderImpl(serviceName, hostName));
     for (int i = 0; i < attemptCount; i++) {
-      hashResult = (serviceName + hostName + i).hashCode();
-      bucketIdx = abs(hashResult) % 100; // hash "test-service"+"lor1-appXXX"
+      when(impl.getHashKey()).thenReturn(serviceName + hostName + i);
```

Comment:
Is there a reason to not use the actual implementation of getHashKey here? Otherwise if it changes the test won't break.

Reply:
fixed. thanks.